src/esp/gfx_batch/Hbao.h

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) Meta Platforms, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#ifndef ESP_GFX_BATCH_HBAO_H_
+#define ESP_GFX_BATCH_HBAO_H_
+
+#include <Corrade/Containers/Pointer.h>
+#include <Magnum/GL/GL.h>
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Vector2.h>
+
+namespace esp {
+namespace gfx_batch {
+
+enum class HbaoFlag {
+  /**
+   * Default should be blur, since without blur there are artifacts. No blur
+   * capability given for debugging.
+   */
+  NoBlur = 1 << 0,
+  UseAoSpecialBlur = 1 << 1,
+  /* These two affect only the cache-aware variant. Only one can be set at
+     most, not both (mutually exclusive) */
+  LayeredImageLoadStore = 1 << 2,
+  LayeredGeometryShader = 1 << 3
+};
+typedef Corrade::Containers::EnumSet<HbaoFlag> HbaoFlags;
+
+CORRADE_ENUMSET_OPERATORS(HbaoFlags)
+
+class HbaoConfiguration {
+ public:
+  Magnum::Vector2i size() const { return size_; }
+
+  HbaoConfiguration& setSize(const Magnum::Vector2i& size) {
+    size_ = size;
+    return *this;
+  }
+
+  HbaoFlags flags() const { return flags_; }
+
+  HbaoConfiguration& setFlags(HbaoFlags flags) {
+    flags_ = flags;
+    return *this;
+  }
+
+  // Blur should always be used, otherwise there's nasty grid-like artifacts.
+  // NoBlur available for debugging.
+  HbaoConfiguration& setNoBlur(bool state) {
+    // special blur is dependent on blur being enabled, so both should be
+    // disabled if NoBlur is enabled.
+    flags_ = (state ? (flags_ | HbaoFlag::NoBlur) & ~HbaoFlag::UseAoSpecialBlur
+                    : flags_ & ~HbaoFlag::NoBlur);
+    return *this;
+  }
+
+  HbaoConfiguration& setUseSpecialBlur(bool state) {
+    // special blur is dependent on blur being enabled, so NoBlur should be
+    // disabled if special blur is enabled
+    flags_ = (state ? (flags_ | HbaoFlag::UseAoSpecialBlur) & ~HbaoFlag::NoBlur
+                    : flags_ & ~HbaoFlag::UseAoSpecialBlur);
+    return *this;
+  }
+
+  HbaoConfiguration& setUseLayeredImageLoadStore(bool state) {
+    // LayeredImageLoadStore and LayeredGeometryShader are mutually exclusive
+    flags_ = (state ? (flags_ | HbaoFlag::LayeredImageLoadStore) &
+                          ~HbaoFlag::LayeredGeometryShader
+                    : flags_ & ~HbaoFlag::LayeredImageLoadStore);
+    return *this;
+  }
+  HbaoConfiguration& setUseLayeredGeometryShader(bool state) {
+    // LayeredImageLoadStore and LayeredGeometryShader are mutually exclusive
+    flags_ = (state ? (flags_ | HbaoFlag::LayeredGeometryShader) &
+                          ~HbaoFlag::LayeredImageLoadStore
+                    : flags_ & ~HbaoFlag::LayeredGeometryShader);
+    return *this;
+  }
```

Comment:
Actually was going to get rid of setFlags(), since we  have dependent/mutually exclusive states that need to be properly tracked. LayeredImage and LayeredGeometry are mutually exclusive, so I don't think they can be subsets of each other.